posts/benchmarking_design_decisions/index.qmd

```diff
@@ -0,0 +1,246 @@
+scale linearly with the number of warnings thrown. Other benchmarking
+experiments
+with `cli` have revealed linear relationships between its speed and the number
+of times it invokes other functions [^3]. The developers of `cli` have also
+conducted benchmarks of the `ansi_` family of functions in `cli` in comparison
+to base R and the `fansi` package. They find that `cli` is consistently slower
+than base R, which corroborates the results of our experiment here.
+Their benchmarks are available in the cli documentation [^4]. So, should we
+be worried about the speed of `cli`? Well, it depends on the context. The
+"R Packages" book by Hadley Wickham and Jenny Bryan has a section on
+this where they suggest approaching such a decision from a holistic, balanced,
+and quantitative approach[^5]. We'll leave the reader to make their own decision
+based on the their use case.
+
+Most developers might argue that this is an optimisation overkill[^6]. However,
+we argue that it is important to consider speed differences in context.
+In the case of simple printing, the speed difference is
+negligible yet disruptive and somewhat painful. However, in the grand scheme
+of things, this might be nothing compared to much slower processes that
+need more attention. In those cases, the developer might want to consider other
+optimisation strategies such as profiling [^7]. The essence of this experiment
+is to demonstrate that benchmarking is one way to help developers make design
+decisions.
+
+## Conclusion
+
+In designing R package infrastructure with dependencies, it might
+sometimes be necessary to test your intuition about their overhead with
+experiments. Here, we have demonstrated how benchmarking is one way to achieve
+this. We show how a simple decision to use `cli::cli_warn()` to handle
+warnings could come at the cost of a tiny loss in speed, which is worth
+considering in its context.
+
+The demonstration here can be extended to other design decisions such as
+input checking, loops, object indexing, and so forth. We recommend benchmarking
+as a way to help developers make design decisions. However, we also recommend
+that developers consider the context of the optimisation in interpreting the results.
+
+## Other R packages for benchmarking
+
+* [microbenchmark](https://github.com/joshuaulrich/microbenchmark): an R
+  package for comparing the execution time of R expressions.
+* [rbenchmark](https://code.google.com/archive/p/rbenchmark/): an R package
+  for benchmarking R code.
+* [tictok](https://github.com/collectivemedia/tictoc): an R package to time
+R functions
+* [touchstone](https://github.com/lorenzwalthert/touchstone/tree/main): an R
+  package for
+  benchmarking of pull requests with statistical confidence.
+
+[^1]: Benchmark (Wikipedia): <https://en.wikipedia.org/wiki/Benchmark_(computing)>
+[^2]:  Hadley Wickham's Advanced R chapter on Exceptions and Debugging: <http://adv-r.had.co.nz/Exceptions-Debugging.html>
```

Comment:
I think this links to the old version of Advanced R. I'm not sure how much the content differs between versions, but it might be worth linking to the newer version of book: https://adv-r.hadley.nz/conditions.html. (Please check the new version contain the relevant info).